src/Api/Controllers/GroupsController.cs

```diff
@@ -38,119 +41,160 @@ public class GroupsController : Controller
+    public async Task<IEnumerable<Guid>> GetUsers(Guid orgId, Guid id)
     {
-        var idGuid = new Guid(id);
-        var group = await _groupRepository.GetByIdAsync(idGuid);
-        if (group == null || !await _currentContext.ManageGroups(group.OrganizationId))
+        var group = await _groupRepository.GetByIdAsync(id);
+        if (group == null)
         {
             throw new NotFoundException();
         }
 
-        var groupIds = await _groupRepository.GetManyUserIdsByIdAsync(idGuid);
+        var authorizationResult =
+            await _authorizationService.AuthorizeAsync(User, group, GroupOperations.Read);
+        if (!authorizationResult.Succeeded)
+        {
+            throw new ResourceAuthorizationFailedException();
+        }
+
+        var groupIds = await _groupRepository.GetManyUserIdsByIdAsync(id);
         return groupIds;
     }
 
     [HttpPost("")]
-    public async Task<GroupResponseModel> Post(string orgId, [FromBody] GroupRequestModel model)
+    public async Task<GroupResponseModel> Post(Guid orgId, [FromBody] GroupRequestModel model)
     {
-        var orgIdGuid = new Guid(orgId);
-        if (!await _currentContext.ManageGroups(orgIdGuid))
+        var organization = await _organizationRepository.GetByIdAsync(orgId);
+        var group = model.ToGroup(orgId);
+
+        var authorizationResult =
+            await _authorizationService.AuthorizeAsync(User, group, GroupOperations.Create);
+        if (!authorizationResult.Succeeded)
         {
-            throw new NotFoundException();
+            throw new ResourceAuthorizationFailedException();
         }
 
-        var organization = await _organizationRepository.GetByIdAsync(orgIdGuid);
-        var group = model.ToGroup(orgIdGuid);
         await _createGroupCommand.CreateGroupAsync(group, organization, model.Collections?.Select(c => c.ToSelectionReadOnly()), model.Users);
 
         return new GroupResponseModel(group);
     }
 
     [HttpPut("{id}")]
     [HttpPost("{id}")]
-    public async Task<GroupResponseModel> Put(string orgId, string id, [FromBody] GroupRequestModel model)
+    public async Task<GroupResponseModel> Put(Guid orgId, Guid id, [FromBody] GroupRequestModel model)
     {
-        var group = await _groupRepository.GetByIdAsync(new Guid(id));
-        if (group == null || !await _currentContext.ManageGroups(group.OrganizationId))
+        var group = await _groupRepository.GetByIdAsync(id);
+        if (group == null)
         {
             throw new NotFoundException();
         }
 
-        var orgIdGuid = new Guid(orgId);
-        var organization = await _organizationRepository.GetByIdAsync(orgIdGuid);
+        var authorizationResult =
+            await _authorizationService.AuthorizeAsync(User, group, GroupOperations.Update);
+        if (!authorizationResult.Succeeded)
+        {
+            throw new ResourceAuthorizationFailedException();
+        }
```

Comment:
This code is duplicated a fair bit in the controllers; do you think it would be feasible to create a [custom attribute](https://learn.microsoft.com/en-us/dotnet/standard/attributes/writing-custom-attributes) for these checks?

Reply:
I thought the same thing but the `AuthorizationHandler` requires the resource object to be passed as a parameter

Reply:
Ah, but! We can create extension methods, check it out: https://github.com/bitwarden/server/pull/2845/commits/639b07be2cc147a86956f67d34c5be5329d64802

Reply:
Update: TIL you can't assert calls to extension methods 😞  so I've created our own auth service wrapper instead - same functionality but more testable.